### ISSUE TITLE
AutoKey process not killed by panel icon context menu Quit option

### ISSUE THREAD

**Issue by Elliria** [autokey/autokey#632]
## Classification:
Bug

## Reproducibility:
Always

## Autokey version:
autokey-gtk 0.96.0-beta.9

Used GUI:
Gtk

Installed via:
pip3

Linux distribution:
- Kubuntu 20.04 LTS running inside of VirtualBox
- KDE Plasma Version 5.18.5
- KDE Frameworks Version: 5.68.0
- Qt Version: 5.12.8
- Kernel Version: 5.11.0-40-generic
- OS Type: 64-bit
- Memory: 3.2 GiB of RAM
- Processor: 1 x Intel Core i5-8600K CPU @ 3.60 GHz

## Summary:
The AutoKey process is not killed by the panel icon's context menu "Quit" option.

## Steps to reproduce:
1. Launch autokey-gtk.
2. Check your system for any autokey processes: ```pgrep -c "autokey"```
    - Expected result: Result should be 1.
    - Actual result: Result is 1.
5. Right-click the AutoKey icon in the panel and choose "Quit".
6. Check your system for any autokey processes: ```pgrep -c "autokey"```
    - Expected result: Result should be 0.
    - Actual result: Result is 1.
 
## Expected result:
Multiple expected results. See above.

## Actual result:
Multiple actual results. See above.

## Notes:
This issue doesn't affect closing AutoKey from the "Quit" option inside the "File" menu in the AutoKey main window. That one properly kills the AutoKey process.

**Comment by BlueDrink9** [comment]
I can't reproduce with gtk. Can you please check again after a restart, and this time first do `pkill autokey; pgrep -c autokey` to ensure there aren't any other processes triggering a false positive?

**Comment by BlueDrink9** [comment]
Also note that autokey does take some time to shut down, so leave 30 seconds in between

**Comment by BlueDrink9** [comment]
I'm going to close for now, but I will re-open if you can confirm it is occuring. Provide logs, too.

**Comment by Elliria** [comment]
I'm attaching a log of the output.

## The steps that I took:
This is exactly what I did to create that log:
1. I booted my VirtualBox virtual machine.
2. I launched AutoKey in a terminal window from a shortcut on my panel.
3. I right-clicked the AutoKey panel icon and chose "Quit" from the menu.
4. I checked for the AutoKey process in another terminal window.
5. The AutoKey process still existed.
6. I waited 30 seconds.
7. I checked for the AutoKey process in another terminal window.
8. The AutoKey process still existed.
9. I waited 3 minutes.
10. I checked for the AutoKey process in another terminal window.
11. The AutoKey process still existed.
12. I did a Ctrl+c keyboard interrupt in the terminal window with AutoKey open in it.
13. I checked for the AutoKey process in another terminal window.
14. The AutoKey process still existed.
15. I waited for about 30 seconds.
16. I checked for the AutoKey process in another terminal window.
17. The AutoKey process still existed.
18. I did another Ctrl-c keyboard interrupt in the window with AutoKey open in it.
19. I checked for the AutoKey process in another terminal window.
20. The AutoKey process no longer existed.

## Additional comments that may prove useful:
* All tests have been and are being done on a clean installation of the most recent Kubuntu LTS that's kept up-to-date and hasn't had any other software added to it other than the AutoKey beta.
* This seems to be an intermittent issue. I can more consistently reproduce it by attempting to open the Qt version of AutoKey (which still won't open for me) before opening the GTK version of AutoKey. I can also more consistently reproduce it by launching AutoKey and then opening any other window so that AutoKey doesn't have focus when I click on its icon in the panel to close it. That said, I can also reproduce it by simply opening and closing the GTK version of AutoKey after a fresh boot without doing anything else first, but it doesn't happen as often.

## The log:
[AutoKeyLog.txt](https://github.com/autokey/autokey/files/7662279/AutoKeyLog.txt)

**Comment by BlueDrink9** [comment]
Is that the log from the steps you described above? It doesn't seem to have any of the wait times you mentioned, but it just might not have captured them. (if you do run another log, please update to beta 10 first because it fixes some of the log clutter)

**Comment by BlueDrink9** [comment]
Also double check that the autokey you are opening is the correct beta - check the about/help menu

**Comment by BlueDrink9** [comment]
Actually, ignore that about beta 10. Having some issues releasing it.

**Comment by Elliria** [comment]
Yes, that's the complete log from all of the steps I described above. It doesn't contain time stamps during those wait times because it's hung or it's stuck or it's wedged or it's in a bad way or it's having a no good very bad day. However you want to look at it, its hands are completely tied and it can't do anything while the clock is ticking. If you do a search for "KeyboardInterrupt" in that log, you'll find the two keyboard interrupts that I did and how AutoKey reacted to the first one and then nothing after the second one because that was success and AutoKey was finally gone.

**Comment by Elliria** [comment]
This is a virtual machine dedicated only to running the AutoKey beta, so the only version of AutoKey that has been installed on it is the **autokey-gtk 0.96.0-beta.9** version as reported above and no other additional software has been or will be installed. I wanted to give it the cleanest possible environment for tests.

**Comment by sebastiansam55** [comment]
Having to press ctrl-c twice to kill autokey when it is running in console is the way it has been for me since I started using the program. 

I don't know if I'd qualify it as expected behavior but it is certainly typical.

**Comment by Elliria** [comment]
Interesting. Pressing **Ctrl+c** normally ends things instantly for me. Anyway, in this case, that's not all I did. I right-clicked the panel icon and chose **Quit** from the context menu. It was only after that failed and after waiting a while for it to succeed (in the event that a delay would help matters) that I pressed **Ctrl+c** twice (with a delay between them) to finally succeed in removing the AutoKey process.

**Comment by BlueDrink9** [comment]
ctrl+c needs two presses, because it isn't specifically handled. I have changed that in `develop`, but it hasn't made its way into the beta.

@Elliria how easy would it be for you to check out/install an old autokey version and test if it works there?

**Comment by Elliria** [comment]
I just did. It's kind of interesting. If I launch the AutoKey beta with just its full path in the latest Kubuntu LTS and then press Ctrl+c twice, AutoKey closes. If I do the same thing with AutoKey 0.90.4 in the previous Ubuntu MATE LTS (not the latest one) and then press Ctrl+c twice, it doesn't close. I pressed it several times and it still didn't close. I had to use the mouse to click the X in the window to close it.

**Comment by BlueDrink9** [comment]
Sorry, I meant testing closing via the tray icon menu!

**Comment by Elliria** [comment]
The tray icon always worked in the old version.

**Comment by Elliria** [comment]
Well, I just did several tests with the new beta (0.96.0-beta.10) and so far, I've only been able to replicate the issue once and it resolved itself in a couple of seconds. In other words, I was able to get a **1** instead of a **0** in response to the ```pgrep -c autokey``` command, but as soon as I ran that pgrep command again a moment later (not even two seconds), I got a **0**. All the other times I've tried to replicate it have been unsuccessful. This is now in a tentative "solved" status, but I'll let you know if I see it again as I continue to mess with the beta.

**Comment by Elliria** [comment]
I spoke too soon. It still has the issue. I triggered it when examining a possibly-related issue in which clicking the X in the upper right of the AutoKey window no longer closes the program, but just closes the main window. I wasn't logging at the time, though, and haven't had a repeat the current issue again since then. I'll get you a new log when I do.

Meanwhile, would you like the steps and log for the other related issue (AutoKey not closing when the X is clicked) in here or would you prefer that as a new issue report?

**Comment by BlueDrink9** [comment]
The new beta shouldn't have changed anything that I know of, so I doubt this is solved

**Comment by BlueDrink9** [comment]
Autokey not closing when the X is clicked sounds like expected behavior. The program shouldn't close just because the window closes, and the X is just from the window manager, managing a window (not a program)

**Comment by Elliria** [comment]
You're right. I don't know what I was thinking. It's not even a change. It's always been like that. Sorry for the noise.

**Comment by BlueDrink9** [comment]
Is this still an issue? And does it *only* apply to the beta?

**Comment by Elliria** [comment]
Yes, it's still an issue. Nothing has changed. No, it doesn't only apply to the beta, but also to AutoKey version 0.95.10 in an identical VM with the only difference in the two being which version of AutoKey is installed.

**Comment by Elliria** [comment]
I thought I'd add that I recently upgraded to first Kubuntu 22.04 LTS and installed AutoKey from the Ubuntu repositories, so I've got AutoKey 0.95.10 running here. I'm having the same intermittent issue with the GTK front-end of AutoKey now that I'm not running it in a VM.

**Symptoms:**

- Sometimes AutoKey instantly quits completely (**pgrep -c autokey** returns 0 when run immediately after using the system tray icon to exit AutoKey.
- Other times, it takes a few seconds for AutoKey to quit completely (**pgrep -c autokey** returns 1 the first few times you run it, but eventually returns 0).
- Other times, AutoKey simply won't quit (**pgrep -c autokey** returns 1 and doesn't change even several minutes later, requiring **pkill autokey** to finish the job).

**Environment:**

- Operating System: Kubuntu 22.04
- KDE Plasma Version: 5.24.4
- KDE Frameworks Version: 5.92.0
- Qt Version: 5.15.3
- Kernel Version: 5.15.0-27-generic (64-bit)
- Graphics Platform: X11
- AutoKey 0.95.10 installed from the Ubuntu repositories

**Comment by josephj11** [comment]
This might need something like strace to see what the holdup is. That's a tool for experts though. It produces a lot of output and you have to know what you're doing to sift through it.

**Comment by Elliria** [comment]
Hopefully an expert will see this and try it. I took a look at strace and it's bigger than me.

**Comment by josephj11** [comment]
Me too.
